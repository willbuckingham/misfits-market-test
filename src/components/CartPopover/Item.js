import React from 'react'
import PropTypes from 'prop-types'
import {
  Typography,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'

import { formatCurrency } from './../../utils'

const useStyles = makeStyles(({ spacing }) => ({
  root: {
    display: 'flex',
  },
  textarea: {
    padding: spacing(1),
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    flexGrow: 1,
  },
  price: {
    display: 'flex',
    alignItems: 'center',
  },
}))

const Item = props => {
  const classes = useStyles()

  const { product, price, quantity } = props

  return (
    <>
      <div className={classes.root}>
        <img src='/fruit.jpg' alt={product} title={product} height={80} />
        <div className={classes.textarea}>
          <Typography>{product}</Typography>
          <Typography color='textSecondary'>{formatCurrency(price)} x {quantity}</Typography>
        </div>
        <div className={classes.price}>
          <Typography variant='h6'>{formatCurrency(price * quantity)}</Typography>
        </div>
      </div>
      <hr/>
    </>
  )
}

Item.propTypes = {
  product: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  quantity: PropTypes.number.isRequired,
}

export default Item
