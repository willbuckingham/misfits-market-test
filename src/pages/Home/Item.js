import React from 'react'
import PropTypes from 'prop-types'
import {
  // IconButton,
  Typography,
  // Button,
  Card,
  CardMedia,
  CardContent,
  CardActions,
  // CardActionArea,
  Fab,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { 
  AddShoppingCart,
} from '@material-ui/icons'

const useStyles = makeStyles(({ spacing }) => ({
  card: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
  },
  cardContent: {
    flexGrow: 2,
  },
  product: {
    lineHeight: 1.2,
  },
  priceContainer: {
    display: 'flex',
    alignItems: 'start',
    justifyContent: 'center',
    marginTop: spacing(1),
  },
  price: {
    fontWeight: 'bold',
  },
  msrp: {
    textDecoration: 'line-through',
  },
}))

export const formatCurrency = v =>
  '$' + v.toFixed(2)

export const calcSavings = (msrp, price) => {
  const savingsExact = price < msrp ? (1 - price / msrp) * 100 : 0
  return Math.round(savingsExact / 5) * 5
}

const Item = props => {
  const classes = useStyles()

  const { id, product, price, msrp, soldOut } = props
  const savings = calcSavings(msrp, price)

  return (
    <Card className={classes.card}>

      <CardMedia
        component='img'
        image='/logo192.png'
        title={product}
        alt={product}
        height={125}
      />
      <CardContent className={classes.cardContent}>
        <Typography
          variant='h6'
          className={classes.product}
        >
          {product}
        </Typography>
        <div className={classes.priceContainer}>
          <Typography display='inline' color='primary' className={classes.price}>{formatCurrency(msrp)}</Typography>
          &nbsp;
          <Typography display='inline' variant='caption' className={classes.msrp} color='textSecondary'>{formatCurrency(price)}</Typography>
          &nbsp;
          { savings && <Typography display='inline' variant='caption' color='textSecondary'>{`${savings}% off MSRP`}</Typography> }
        </div>
      </CardContent>
      <CardActions>
        { soldOut ?
          <Typography color='error'>Sold Out!</Typography> :
          <Fab color='secondary' size='small'>
            <AddShoppingCart/>
          </Fab>
        }
      </CardActions>
    </Card>
  )
}

Item.propTypes = {
  id: PropTypes.number.isRequired,
  product: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  msrp: PropTypes.number.isRequired,
  soldOut: PropTypes.bool.isRequired,
}

export default Item
