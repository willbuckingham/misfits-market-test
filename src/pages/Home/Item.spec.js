import React from 'react'
import { shallow } from 'enzyme'

import Item, { formatCurrency, calcSavings } from './Item'

describe('Item', () => {
  let props

  beforeAll(() => {
    props = {
      id: 1,
      product: 'Product',
      price: 1,
      msrp: 2,
      soldOut: false,
      onAdd: jest.fn(),
    }
  })

  it('renders', () => {
    expect(shallow(<Item { ...props } />)).toMatchSnapshot()
  })

  it('renders sold out', () => {
    expect(shallow(<Item { ...props } soldOut={true} />)).toMatchSnapshot()
  })
})

describe('formatCurrency', () => {
  it('should format', () => {
    expect(formatCurrency(0)).toBe('$0.00')
    expect(formatCurrency(1)).toBe('$1.00')
    expect(formatCurrency(1.5)).toBe('$1.50')
    expect(formatCurrency(1.55)).toBe('$1.55')
    expect(formatCurrency(1.555)).toBe('$1.55')
  })
})

describe('calcSavings', () => {
  it('is equal or more expensive', () => {
    expect(calcSavings(5, 5)).toBe(0)
    expect(calcSavings(4, 5)).toBe(0)
  })

  it('should round to 20', () => {
    expect(calcSavings(5, 4)).toBe(20)
  })
})
