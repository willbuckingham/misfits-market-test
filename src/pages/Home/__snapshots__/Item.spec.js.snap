// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-card-1"
>
  <WithStyles(ForwardRef(CardMedia))
    alt="Product"
    component="img"
    height={125}
    image="/logo192.png"
    title="Product"
  />
  <WithStyles(ForwardRef(CardContent))
    className="makeStyles-cardContent-2"
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-product-3"
      variant="h6"
    >
      Product
    </WithStyles(ForwardRef(Typography))>
    <div
      className="makeStyles-priceContainer-4"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-price-5"
        color="primary"
        display="inline"
      >
        $2.00
      </WithStyles(ForwardRef(Typography))>
       
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-msrp-6"
        color="textSecondary"
        display="inline"
        variant="caption"
      >
        $1.00
      </WithStyles(ForwardRef(Typography))>
       
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
        display="inline"
        variant="caption"
      >
        50% off MSRP
      </WithStyles(ForwardRef(Typography))>
    </div>
  </WithStyles(ForwardRef(CardContent))>
  <WithStyles(ForwardRef(CardActions))>
    <WithStyles(ForwardRef(Fab))
      color="secondary"
      size="small"
    >
      <AddShoppingCartIcon />
    </WithStyles(ForwardRef(Fab))>
  </WithStyles(ForwardRef(CardActions))>
</WithStyles(ForwardRef(Card))>
`;

exports[`Item renders sold out 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-card-1"
>
  <WithStyles(ForwardRef(CardMedia))
    alt="Product"
    component="img"
    height={125}
    image="/logo192.png"
    title="Product"
  />
  <WithStyles(ForwardRef(CardContent))
    className="makeStyles-cardContent-2"
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-product-3"
      variant="h6"
    >
      Product
    </WithStyles(ForwardRef(Typography))>
    <div
      className="makeStyles-priceContainer-4"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-price-5"
        color="primary"
        display="inline"
      >
        $2.00
      </WithStyles(ForwardRef(Typography))>
       
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-msrp-6"
        color="textSecondary"
        display="inline"
        variant="caption"
      >
        $1.00
      </WithStyles(ForwardRef(Typography))>
       
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
        display="inline"
        variant="caption"
      >
        50% off MSRP
      </WithStyles(ForwardRef(Typography))>
    </div>
  </WithStyles(ForwardRef(CardContent))>
  <WithStyles(ForwardRef(CardActions))>
    <WithStyles(ForwardRef(Typography))
      color="error"
    >
      Sold Out!
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(CardActions))>
</WithStyles(ForwardRef(Card))>
`;
